### DISCUSSION TITLE
micropython/lora, reliable/delivery TX power adjust

### DISCUSSION THREAD

**Discussion by davefes** [micropython/micropython#12975] [Libraries & Drivers] [answered]
Is the normal procedure to set TX power at the receiver end to the max allowed and let the sensor or sender end do any power adjustment?

Just trying to think through a bi-directional link and how two transmitters would behave.  Should one just disable TX power adjustment at the end where you have the most free power, ie mains-connected?

**Comment by kjm1102** [comment, marked as answer]
After 8 years of trying to make sense of RF signal strengths with lora, wifi, LTE, & LEOs I've decided they're OK for window dressing but useless for much else. You really need to check RSSIs in concert with SNRs to have any chance of making an informed decision about appropriate  transmit power & even then there is a lot of guesswork on the weightings to apply to each.  

Furthermore RF circumstances are often localised & jittery so guesstimating reply Tx pwr should really be based on rssi/snr levels at the target site rather than the values where you are sending from. I never cease to be amazed how asymmetric these readings can be between sites with identical boards/antennas.   

My advice is to manage Tx pwr based on BERs or some similar indicator of how hard a link is actually working rather than airy fairy RF parameters.

**Reply by davefes** [reply]
In Wei's LoRa implementation https://github.com/Wei1234c/SX127x_driver_for_MicroPython_on_ESP8266/tree/master he measures both RSSI and SNR at the target end. I set code in those links that when SNR went to zero, which always seemed to be -120dBm I would flag it as a fail.  But, I think there still some fails even at -90 to -100dBm

In micropython/lora the default threshold S/W will increase TX power to try and keep the RSSI above -110dBm.  At the level I still the odd error:
- Duplicate packet received from 0x1fa
- Invalid checksum. calc=0x8a received=0x8f
- Invalid length in payload header
  Skipped/lost 1 packets from 0x1fa

I have quite a bit of experience testing digital radio links (iPerf), but very little understanding of how LoRa modulation behaves in noisy, multi-path, influence of other RF sources and etc. 

Thanks for your comments.

**Reply by kjm1102** [reply]
Seems like a nice thorough driver as far as diagnostics go. I've been meaning to check out the lora baked into the latest upython. Read about it on a discord somewhere & have been meaning to try it.

**Reply by davefes** [reply]
> the lora baked into the latest upython

Has something happened since this comment:
https://github.com/orgs/micropython/discussions/10807

**Comment by davefes** [comment]
Further to your comments about RSSI and SNR and why you might get different RSSIs at each end.  I suggest that the test environment could have a big impact.  Years ago, even though I had LOS to a cell tower about 2KMs away I was not able to make a phone call.  The only think I could think of was multi-path distortion.  So, I hit Google <multi-path Lora> and the first link, if you like "light-reading".   

https://www.google.com/url?sa=t&rct=j&q=&esrc=s&source=web&cd=&cad=rja&uact=8&ved=2ahUKEwintIONh8eCAxX2TmwGHRZhBvgQFnoECCIQAQ&url=https%3A%2F%2Farxiv.org%2Fpdf%2F2201.03825&usg=AOvVaw22lyQz34uaNZkebcPiPPH-&opi=89978449

https://medium.com/@patburns/lora-in-multipath-theory-vs-experiment-fae2ccd10242
mentions Rayleigh flat fading and how Haystack XR2 improves things. 

I modified the Reliable_delivery example to do a crude packet test.  I have got 1000 packets so far with no errors at RSSI -108dBm.  Things seem to start falling-over at -109/-110dBm  (bw = 125kHz, sf = 10)

Just need someone to port Haystack XR2 to Micropython!

**Reply by kjm1102** [reply]
Interesting take on light reading you have there Dave. I'm currently half way through Dave Lindley's 'The Dream Universe' which has a bit to say about the way we try to understand things & how easy it is to overdo the math.

Twenty to thirty dB improvement for XR2 is a bold claim but not many XR2 hits online post 2020. I agree it would be nice to have a stack to try it out on a upython board.

Looks like my memory of the discord chatter was wrong https://github.com/micropython/micropython-lib/tree/master/micropython/lora has to be loaded up manually. They've gone the full porky, so many bells & whistles, I'm gonna stick with the modified Lemariva for now, the TTGOs are working down to -129dBm, if it ain't broke don't fix it.

Roger your comments on cellular. I have a dead spot outside my office window that none of the IoT modems I've tried can work from to a tower just a few km away. All the slow data schemes (nbiot, catm1, etc) are on B28/750MHz in AU which should be a nice frequency diffraction wise but the lack of diversity antenna capability on the boards is the problem I think. Performance wise the results are bizarre. I've got sites where the rsrp sits @ ~-105dBm that don't miss a beat & other locations where only half the uploads make it to the server @ -75dbm nominal.

**Reply by davefes** [reply]
What LoRa parameters did you use to get to -129dBm?  I imagine that is with some packet loss. 

micropython/lora is baked into the latest Nucleo WB55.  I have some CEByte E77-400M22S modules coming but looks like I might have to re-visit `mpy-cross` to get it to work.  Nice looking module and about 1/2 the price of a WeACT BlackPill plus a Hope Radio RFM96W.

**Reply by kjm1102** [reply]
bw=125k, sf=12. Yeah that's the limit I never see a -130dBm on a good decode.
I think I looked at the E77s once, they have the TCXO so you can use BWs<125k? I squibbed it though, still traumatised from having to change from pycom boards/code to what I'm on now. Having to redo code from one flavour of upython to another is misery on a stick.

**Reply by davefes** [reply]
The E77-400M22S uses a 32MHz xtal, don't see TCXO mentioned in the datasheet.
